src/PairedRouting.php

```diff
@@ -772,6 +772,45 @@ public function add_permalink_settings_notice() {
 		<?php
 	}
 
+	/**
+	 * Update canonical URL for legacy reader paired urls.
+	 *
+	 * @param string $redirect_url Canonical redirect URL.
+	 *
+	 * @return string Updated canonical URL.
+	 */
+	public function maybe_update_paired_url( $redirect_url ) {
+
+		global $wp_rewrite;
+
+		/**
+		 * To address issue AMP version of a comment page.
+		 * Reference: https://github.com/ampproject/amp-wp/issues/6355
+		 */
+		if ( ! empty( $redirect_url ) &&
+			! empty( $wp_rewrite->comments_pagination_base ) &&
+			AMP_Options_Manager::get_option( Option::PAIRED_URL_STRUCTURE ) === Option::PAIRED_URL_STRUCTURE_LEGACY_READER
+		) {
+
+			$amp_slug = amp_get_slug();
+			$amp_slug = ( ! empty( $amp_slug ) ) ? $amp_slug : 'amp';
+
+			/**
+			 * Reference: https://regex101.com/r/544758/1/
+			 */
+			$regex = sprintf(
+				":/(%s-[0-9]+?)/%s/\\1(/+)?($|\?|\#):",
+				preg_quote( $wp_rewrite->comments_pagination_base, ':' ),
+				preg_quote( $amp_slug, ':' )
+			);
```

Comment:
`amp_get_slug()` returns `'amp'` by default, but if it did not you could use the `QueryVar::AMP` constant.

I don't think we need to document the reference to the regex either (any thoughts on this @westonruter?)

```suggestion
			$regex = sprintf(
				":/(%s-[0-9]+?)/%s/\\1(/+)?($|\?|\#):",
				preg_quote( $wp_rewrite->comments_pagination_base, ':' ),
				preg_quote( amp_get_slug(), ':' )
			);
```

Reply:
Suggestion looks good.